Make named parameters work with :name, $name, @name and name

haaawk opened the pull request:
Fixes #237

haaawk commented on sqld/src/query.rs:
```diff
@@ -381,8 +381,8 @@ mod test {
         let mut params = HashMap::new();
-        params.insert(":first".to_owned(), Value::Integer(10));
-        params.insert("$second".to_owned(), Value::Integer(20));
+        params.insert("first".to_owned(), Value::Integer(10));
+        params.insert("second".to_owned(), Value::Integer(20));
```
😄

MarinPostma commented on sqld/src/query.rs:
```diff
@@ -389,6 +391,19 @@ mod test {
+    #[test]
+    fn test_bind_params_positional_named_no_prefix() {
+        let con = rusqlite::Connection::open_in_memory().unwrap();
+        let mut stmt = con.prepare("SELECT :first || $second").unwrap();
```
What happens if you try with a statement like 'select :first || $first'?

haaawk replied:
It won't allow it:
```
missing parameters, expected 2 found 1
```
here's the change:
```
diff --git a/sqld/src/query.rs b/sqld/src/query.rs
index b75bdf5..739f689 100644
--- a/sqld/src/query.rs
+++ b/sqld/src/query.rs
@@ -394,10 +394,9 @@ mod test {
     #[test]
     fn test_bind_params_positional_named_no_prefix() {
         let con = rusqlite::Connection::open_in_memory().unwrap();
-        let mut stmt = con.prepare("SELECT :first || $second").unwrap();
+        let mut stmt = con.prepare("SELECT :first || $first").unwrap();
         let mut params = HashMap::new();
         params.insert("first".to_owned(), Value::Integer(10));
-        params.insert("second".to_owned(), Value::Integer(20));
         let params = Params::new_named(params);
         params.bind(&mut stmt).unwrap();
```

haaawk replied:
When I worked around it by adding second parameter `$first`, it failed:
```
```
here's the change:
```
```

haaawk replied:
Thanks for pointing it out @MarinPostma

haaawk replied:
done.

haaawk commented:
I swapped the order of fetching named params so that $name, :name and @name will take precedence over name.
Added a test to verify that.

MarinPostma reviewed (commented):
can you also update this document to reflect the changes? https://github.com/libsql/sqld/blob/main/docs/http_api.md

haaawk commented:
I rebased, changed the check that number of params in statement <= number of params in request to checking that each param in statement has a match in request and updated the docs.

MarinPostma reviewed (approved):
looks good :)

bors merge

MarinPostma commented:
bors merge
